signer/src/ecdsa.rs

```diff
@@ -8,85 +8,128 @@
+//! - **SignECDSA**: A trait implemented by data types that can be signed using ECDSA.
 //!
-//! ## Examples
-//!
-//! ### Signing and Verifying a String
-//!
-//! The following example demonstrates how to sign a simple string and then verify its signature:
-//!
-//! ```
-//! use sha2::Digest;
-//! use signer::ecdsa::SignEcdsa;
-//! use signer::keys::PrivateKey;
-//!
-//! struct SignableStr(&'static str);
-//!
-//! // Implementing `ProtoSerializable` and conversion traits unlock the signing
-//! // functionality in this module.
-//! #[allow(clippy::derive_partial_eq_without_eq)]
-//! #[derive(Clone, PartialEq, ::prost::Message)]
-//! pub struct ProtoSignableStr {
-//!     /// The string
-//!     #[prost(string, tag = "1")]
-//!     pub string: ::prost::alloc::string::String,
-//! }
-//!
-//! impl ProtoSerializable for SignableStr {
-//!     type Message = ProtoSignableStr;
-//!
-//!     fn type_tag(&self) -> &'static str {
-//!         "SBTC_SIGNABLE_STR"
-//!     }
-//! }
-//!
-//! let msg = SignableStr("Sign me please!");
-//! let private_key = PrivateKey::try_from(&p256k1::scalar::Scalar::from(1337)).unwrap();
-//!
-//! // Sign the message.
-//! let signed_msg = msg.sign_ecdsa(&private_key);
-//!
-//! // Verify the signed message.
-//! assert_eq!(signed_msg.recover_ecdsa(), Ok(private_key.into()));
 
-use secp256k1::SECP256K1;
+use prost::bytes::Buf as _;
+use prost::Message as _;
+use secp256k1::ecdsa::RecoverableSignature;
+use sha2::Digest as _;
 
 use crate::codec::ProtoSerializable;
+use crate::error::Error;
 use crate::keys::PrivateKey;
 use crate::keys::PublicKey;
-use crate::signature::SighashDigest as _;
+use crate::message::SignerMessage;
+use crate::proto;
+use crate::signature::RecoverableEcdsaSignature;
 
 /// Wraps an inner type with a public key and a signature,
 /// allowing easy verification of the integrity of the inner data.
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub struct Signed<T> {
     /// The signed structure.
     pub inner: T,
-    /// The public key of the signer.
-    pub signer_pub_key: PublicKey,
     /// A signature over the hash of the inner structure.
-    pub signature: Vec<u8>,
+    pub signature: RecoverableSignature,
 }
 
-impl<T> Signed<T>
-where
-    T: ProtoSerializable + Clone,
-    T: Into<<T as ProtoSerializable>::Message>,
-{
-    /// Verify the signature over the inner data.
-    pub fn verify(&self) -> bool {
-        let msg = secp256k1::Message::from_digest(self.inner.digest());
-        let Ok(sig) = secp256k1::ecdsa::Signature::from_compact(&self.signature) else {
-            return false;
-        };
+impl Signed<SignerMessage> {
+    /// Determines the public key for which `sig` is a valid signature for
+    /// `msg`.
+    pub fn recover_ecdsa(&self) -> Result<PublicKey, Error> {
+        let msg = secp256k1::Message::from_digest(self.inner.to_digest());
+        self.signature.recover_ecdsa(&msg)
```

Comment:
nit: reuse `recover_ecdsa_with_digest` with `self.inner.to_digest()`

Reply:
Done in https://github.com/stacks-network/sbtc/pull/958/commits/3078ca8c7a44e4526cbf15dbc82c66f3b6e67416. But I actually need to remove that function entirely.